geo/src/algorithm/affineops.rs

```diff
@@ -0,0 +1,287 @@
+use crate::{CoordFloat, CoordNum, Coordinate, Point};
+use std::fmt;
+
+/// A general affine transformation matrix
+///
+/// Note that affine transforms using the same numeric type (e.g. [`CoordFloat`](crate::CoordFloat)) can be **composed**,
+/// and the result can be applied to geometries using e.g. [`MapCoords`](crate::MapCoords)
+///
+/// We conceive of `AffineTransform` as a row-major matrix.
+/// 2D affine transforms require six matrix parameters:
+///
+/// `[a, b, d, e, xoff, yoff]`
+///
+/// these map onto the `AffineTransform` rows as follows:
+///
+/// `[[a, b, xoff],`
+///
+/// `[d, e, yoff],`
+///
+/// `[0, 0, 1]]`
+///
+/// The equations for transforming coordinates `(x, y) -> (x', y')` are given as follows:
+///
+/// `x' = ax + by + xoff`
+///
+/// `y' = dx + ey + yoff`
+/// # Examples
+///
+/// ```
+/// use geo::{MapCoords, AffineTransform, AffineOps, AffineSkew};
+/// use geo::{line_string, BoundingRect, point, LineString, Centroid};
+///
+/// let ls: LineString<f64> = line_string![
+///     (x: 0.0f64, y: 0.0f64),
+///     (x: 0.0f64, y: 10.0f64),
+/// ];
+/// let origin = ls.bounding_rect().unwrap().centroid();
+/// let op = AffineTransform::identity().skewed(45.0, 45.0, origin);
+/// let skewed = ls.map_coords(|coord| op.apply(coord));
+///
+/// assert_eq!(skewed, line_string![
+///     (x: -4.999999999999999f64, y: 0.0f64),
+///     (x: 4.999999999999999f64, y: 10.0f64)
+/// ]);
+/// ```
+#[derive(Copy, Clone)]
+pub struct AffineTransform<T: CoordNum>([[T; 3]; 3]);
+
+impl<T: CoordNum> Default for AffineTransform<T> {
+    fn default() -> Self {
+        AffineOps::identity()
+    }
+}
+
+impl<T: CoordNum> fmt::Debug for AffineTransform<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("AffineTransform")
+            .field("a", &self.0[0][0])
+            .field("b", &self.0[0][1])
+            .field("xoff", &self.0[1][2])
+            .field("d", &self.0[1][0])
+            .field("e", &self.0[1][1])
+            .field("yoff", &self.0[1][2])
+            .finish()
+    }
+}
+
+/// Affine operations common to all numeric types
+///
+/// These operations are available for both integer and floating point types.
+/// The two Affine transforms that currently only work with floating point types are defined in
+/// [`AffineRotate`](AffineRotate), and [`AffineSkew`](AffineSkew).
+pub trait AffineOps<T: CoordNum> {
```

Comment:
I just realized I was a little confused when I first read this. I had assumed AffineOps would add methods to Geometry types - but I'm realizing it's only implemented for AffineTransform.

Is there a reason that the `AffineOps` trait exists at all at this point, vs having the implementations directly on AffineTransform? 

I'm having a hard time imagining what else would ever implement `AffineOps` besides `AffineTransform`.

Reply:
Possibly because I confused myself when defining the supertrait bounds on `AffineSkew` and `AffineRotate` – maybe there's no need for the supertrait complication if I can define them directly on `AffineTransform`. In my (flimsy) defence my brain had melted after six hours of traits. Let me try it out.

Reply:
All `CoordNum` methods are now directly re-implemented on `AffineTransform`, with the two `CoordFloat` ops as traits. I don't _think_ there's any way to directly impl them?

Reply:
👍👍👍 works perfectly